pkg/actuators/machine/instances_test.go

```diff
@@ -15,3 +15,1242 @@ limitations under the License.
+							ImageId: stubImageID,
+							Status:  EcsImageStatusAvailable,
+						},
+					},
+				},
+			},
+			instancesRequest: createDescribeInstancesRequest(getInstanceIds()),
+			instancesResponse: &ecs.DescribeInstancesResponse{
+				Instances: ecs.InstancesInDescribeInstances{
+					Instance: []ecs.Instance{},
+				},
+			},
+			runInstancesResponse: &ecs.RunInstancesResponse{
+				InstanceIdSets: ecs.InstanceIdSets{
+					InstanceIdSet: []string{stubInstanceID},
+				},
+			},
+			runInstancesRequest: &ecs.RunInstancesRequest{
+				ImageId:         stubImageID,
+				InstanceType:    stubInstanceType,
+				SecurityGroupId: stubSecurityGroupID,
+				VSwitchId:       stubVSwitchID,
+				Tag:             covertToRunInstancesTag(stubTagList),
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockAlibabaCloudClient := mock.NewMockClient(mockCtrl)
+
+			mockAlibabaCloudClient.EXPECT().DescribeImages(gomock.Any()).Return(tc.imageResponse, tc.imageErr).AnyTimes()
+			mockAlibabaCloudClient.EXPECT().ListResourceGroups(gomock.Any()).Return(tc.resourceGroupsResponse, tc.resourceGroupesErr).AnyTimes()
+			mockAlibabaCloudClient.EXPECT().DescribeSecurityGroups(gomock.Any()).Return(tc.securityGroupResponse, tc.securityGroupErr).AnyTimes()
+			mockAlibabaCloudClient.EXPECT().DescribeVSwitches(gomock.Any()).Return(tc.vswitchesResponse, tc.vswitchesErr).AnyTimes()
+			mockAlibabaCloudClient.EXPECT().RunInstances(gomock.Any()).Return(tc.runInstancesResponse, tc.runInstancesErr).AnyTimes()
+			mockAlibabaCloudClient.EXPECT().DescribeInstances(gomock.Any()).Return(tc.instancesResponse, tc.instancesErr).AnyTimes()
+
+			_, runErr := runInstances(machine, tc.providerConfig, "", mockAlibabaCloudClient)
+			t.Log(runErr)
+			if runErr == nil {
+				if !tc.succeeds {
+					t.Errorf("Call to runInstances did not fail as expected")
+				}
+			} else {
+				if tc.succeeds {
+					t.Errorf("Call to runInstances did not succeed as expected")
+				}
+			}
+		})
+	}
+}
+
+func Test_buildDescribeSecurityGroupsTag(t *testing.T) {
+	cases := []struct {
+		tagList  []machinev1.Tag
+		expected []ecs.DescribeSecurityGroupsTag
+	}{
+		{
+			// empty tags
+			tagList:  []machinev1.Tag{},
+			expected: []ecs.DescribeSecurityGroupsTag{},
+		},
+		{
+			tagList: []machinev1.Tag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+			},
+			expected: []ecs.DescribeSecurityGroupsTag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+			},
+		},
+		{
+			// multiple duplicate tags
+			tagList: []machinev1.Tag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
+			},
+			expected: []ecs.DescribeSecurityGroupsTag{
+				{Key: "clusterID", Value: "test-ClusterID"},
+				{Key: "clusterSize", Value: "test-ClusterSize"},
+				{Key: "clusterSize", Value: "test-ClusterSizeDuplicatedValue"},
```

Comment:
Are we not expecting this function to filter out duplicate values?

Reply:
i fixed func to filter duplicate key.